File: data-structure-optimizing-pereformance/my-app/src/main/java/com/problems/TwoSum.java
package com.problems;

import java.util.HashMap;

public class TwoSum {
    
    // Solution 1
    // Time Complexity: O(n^2)
    //  Space Complexity: O(1)
    public static int[] TwoSum(int[] nums, int target){
        int length = nums.length ; 
        for (int i=0 ; i < length-1; i++){
            for ( int j=i+1 ; j < length ; j++){
                
                if (nums[i] + nums[j] == target){
                    int[] indexes = new int[2]; 
                    indexes[0] = i;
                    indexes[1] = j;
                    return indexes; 
                }
            }
        }
        return null ; 
    }

    /*  Solution 2
        Time Complexity: O(n)
        Space Complexity: O(n)
     */
     
    public static int[] TwoSumV(int[] nums, int target){
        HashMap<Integer, Integer> indexMap = new HashMap<>() ; 
        for (int i=0 ; i<nums.length ; i++){
            Integer requiredNumber = target - nums[i] ;
            System.out.println(requiredNumber +  " " + nums[i]);            
            if (indexMap.containsKey(requiredNumber)){
                int indexes[] = {indexMap.get(requiredNumber), i} ; 
                return indexes;
            }
            indexMap.put((nums[i]), i) ; 
        }
        return null ;
    }
    
    public static void main(String[] args) {
        int[] nums = { 7, 2, 13 ,11 } ; 
        int[] indexes = TwoSumV(nums , 9) ;
         System.out.println("[ " +  indexes[0] + ", " + indexes[1] + " ]"); 

    }
}

File: leetcode/random/two_sum.py

# time o(n)
# space o(n)
# hashmap
def two_sum_indexes(nums , target):
    inx_map = {}
    for i in  range(len(nums)) :
        required_number = target - nums[i]
        if required_number in inx_map:
            return [i, inx_map.get(required_number)]
        inx_map[nums[i]] = i 

# time  O(n)
# space O(n)
# using set 
def is_two_sum(nums , target):
    reuired = set()
    for i in  range(len(nums)) :
        required_number = target - nums[i]
        if required_number in reuired:
            return True
        reuired.add(nums[i])
    
#  o(n^2)
def twoSumt(nums , target):
    last = len(nums) - 1  
    i = 0
    while i < last :
        if nums[last] + nums[i] == target:
            return [i, last]
        i += 1
        if i == last:
            last -= 1
            i= 0

# o(n^n)
def two_sum(nums, target):
    for i in range(len(nums)):
        for j in range(i+1, len(nums)):
            if nums[j] + nums[i] == target:
                return [j, i]
    

######### 

# nums = [2, 7, 11 ,15]
nums = [1, 2, 4 ,4]
target= 8
print(two_sum_indexes(nums, target))
